### PR TITLE
Replace rmm::mr::device_memory_resource* with rmm::device_async_resource_ref

### LINKED ISSUE #1371: Replace device_memory_resource* with device_async_resource_ref
Update occurences of `rmm::mr::device_memory_resource*` to
`rmm::device_async_resource_ref` in this repo. The rationale is explained
in [this document](https://docs.google.com/document/d/1QsLQn17qexczFGbzjLzTkIiGztfaQvLu-DpNu2EPw1E/edit#heading=h.wdftxmu1lxt).

### PR DESCRIPTION
**Pull request by harrism**
## Description

Closes #1371

Replaces all occurrences of `rmm::mr::device_memory_resource*` in parameters with `rmm::device_async_resource_ref`.  Also updates guidance in the DEVELOPER_GUIDE.md

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Review by trxcllnt — APPROVED** [review]

**Review by isVoid — APPROVED** [review]

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -245,6 +246,27 @@ RMM provides a "default" memory resource for each device that can be accessed an
 respectively. All memory resource parameters should be defaulted to use the return value of
 `rmm::mr::get_current_device_resource()`.
 
+### Resource Refs
+
+Memory resources are passed via resource ref parameters. A resource ref is memory resource wrapper
+that enables consumers to specify properties of resources that they expect. These are defined
+in the `cuda::mr` namespace of libcu++, but RMM provides some convenience wrappers in
+`rmm/resource_ref.hpp`:
+ - `rmm::device_async_resource_ref` accepts a memory resource that provides synchronous allocation
```
Comment (isVoid) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:255]: Do you mean `device_resource_ref` here? It looks like a duplicate of below.

Comment (harrism) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:255]: Yes typo. Made same error in libcudf.

Comment (harrism) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md]: Fixed

**Pull request merged**